src/exoplanet_core/pymc/ops.py

```diff
@@ -0,0 +1,235 @@
+            of ``b`` or ``r``.
+
+    """
+    __props__ = ()
+
+    def make_node(self, b, r):
+        in_args = [as_tensor_variable(b), as_tensor_variable(r)]
+        if any(i.dtype != "float64" for i in in_args):
+            raise ValueError("float64 precision is required")
+        x = in_args[0]
+        o = [
+            pt.tensor(
+                # NOTE: Changed broadcastable to shape because it caused deprecation warning,
+                # BUT this might change again in the future: https://github.com/pymc-devs/pytensor/issues/408
+                shape=tuple(x.broadcastable) + (False,),
+                dtype=x.dtype,
+            )
+            for _ in range(3)
+        ]
+        return basic.Apply(self, in_args, o)
+
+    def infer_shape(self, *args):
+        shapes = args[-1]
+        shape = tuple(shapes[0]) + (3,)
+        return shape, shape, shape
+
+    def perform(self, node, inputs, outputs):
+        b, r = inputs
+        shape = b.shape + (3,)
+        s = resize_or_set(outputs, 0, shape)
+        dsdb = resize_or_set(outputs, 1, shape)
+        dsdr = resize_or_set(outputs, 2, shape)
+        driver.quad_solution_vector_with_grad(b, r, s, dsdb, dsdr)
+
+    def grad(self, inputs, gradients):
+        b, r = inputs
+        s, dsdb, dsdr = self(b, r)
+        bs = gradients[0]
+
+        for g in gradients[1:]:
+            if not isinstance(g.type, pytensor.gradient.DisconnectedType):
+                raise ValueError(
+                    "Backpropagation is only supported for the solution vector"
+                )
+
+        if isinstance(bs.type, pytensor.gradient.DisconnectedType):
+            return [
+                pytensor.gradient.DisconnectedType()(),
+                pytensor.gradient.DisconnectedType()(),
+            ]
+
+        return [pt.sum(bs * dsdb, axis=-1), pt.sum(bs * dsdr, axis=-1)]
+
+    def R_op(self, inputs, eval_points):
+        if eval_points[0] is None:
+            return eval_points
+        return self.grad(inputs, eval_points)
+
+
+_quad_solution_vector = QuadSolutionVector()
+
+
+def quad_solution_vector(b, r):
+    return _quad_solution_vector(b, r)[0]
+
+
+# ******************
+# * CONTACT POINTS *
+# ******************
+class ContactPoints(op.Op):
+    __props__ = ()
+
+    def make_node(self, a, e, cosw, sinw, cosi, sini, L):
+        in_args = list(
+            map(as_tensor_variable, (a, e, cosw, sinw, cosi, sini, L))
+        )
+        if any(i.dtype != "float64" for i in in_args):
+            raise ValueError("float64 precision is required")
+        out_args = [
+            in_args[0].type(),
+            in_args[0].type(),
+            pt.tensor(
+                # NOTE: Changed broadcastable to shape because it caused deprecation warning,
+                # BUT this might change again in the future: https://github.com/pymc-devs/pytensor/issues/408
+                shape=tuple(in_args[0].broadcastable),
```

Comment:
Should this be something like

```suggestion
                shape=tuple(in_args[0].shape),
```

instead? I haven't been following the pytensor shape discussions.

Reply:
After looking a bit more into this, I see two options:

- What I did above, using `x.broadcastable` directly. Internally, PyTensor converts the boolean values from `broadcastable` to either `1` or `None` to put them in the object's shape. It's the most flexible option, and the closest to the previous behaviour (and the current one in the PyMC 3 backend)
- The other option is to assign `shape` to an actual shape through `x.type.shape`, and append `3` to it.

The first one has the advantage of being closer to the PyMC3 implementation. The second one is a bit more explicit and constraining, and (maybe?) a bit more stable against potential future changes (see https://github.com/pymc-devs/pytensor/issues/408, I subscribed to the issue to know if there are breaking changes in any case). I implemented the first one for now, with comments, but it's easy to switch to the second one. Let me know what you think!

(Using `x.shape` directly does not work because it returns a PyTensor `Shape` object instead of a tuple with booleans or numbers (`x.type.shape` does that))